### ISSUE TITLE
Do you have detailed documents on the execution kernel (Device Runner, for example)?

### ISSUE THREAD

**Issue by rogerzico** [rwth-i6/returnn#38]
As I explore deeper into this application for debugging and other purposes, I thought it would be nice to learn how you guys architect the task running component.

Thanks!

**Comment by albertz** [comment]
I don't exactly understand what you mean by "task running component".
Have you looked at the documentation ([here](http://returnn.readthedocs.io/en/latest/))?
Have you looked into the code itself? There are lots of comments.
Please be more specific what you think should be documented better. Which part in the code, in what file, in what function is unclear?

**Comment by rogerzico** [comment]
That's fine; I can get by probing the code:

----------------
EngineTask.py:

class TaskThread:

      def device_run(self):
        batch_idx = self.run_start_batch_idx = self.devices_batches_idx
        assert len(self.alloc_devices) == len(self.allocated_devices_batches)
        self.running_devices_batches = self.allocated_devices_batches
        for device, batches in zip(self.alloc_devices, self.running_devices_batches):
       ........
          print("on device", device.name, file=log.v5)
          **device.run(self.parent.task)**
---------------

I guess every network layer's computation graph node is tied to "task" !

**Comment by albertz** [comment]
See `Device.py`. It creates two network computation graphs `trainnet` and `testnet`, one for training (e.g. Dropout etc enabled) and one for testing (Dropout disabled). A `Device` instance is per process/GPU/device.

Note that this is for Theano, which is deprecated. The TensorFlow backend has a separate implementation, and has much more documentation in the code.

I'm closing this issue now as it is not really a bug.
